### PR TITLE
[X86][GlobalISel] Enable G_SDIV/G_UDIV/G_SREM/G_UREM

### PR DESCRIPTION
**Pull request by e-kud**
* Create a libcall for s64 type for 32 bit targets.
* Fix a bug in REM selection: SUBREG_TO_REG is not intended to produce a value from super registers.
* Replace selector tests by end-to-end tests. Other passes check the selected MIR better.

### CODE REVIEW

**Comment by tschuett** [comment]
One of the nice features of GlobalIsel is that you can test passes in isolation. Why did the mir tests did not reveal the bug?
End-to-end tests are a SelectionDag feature.

**Comment by arsenm** [comment]
> One of the nice features of GlobalIsel is that you can test passes in isolation. Why did the mir tests did not reveal the bug? End-to-end tests are a SelectionDag feature.

Neither is a replacement for each other. I would add new end to end tests, not delete existing MIR tests

**Comment by tschuett** [comment]
I would also amend the select mir files to show the fixed bug.

**Comment by topperc** [comment]
> One of the nice features of GlobalIsel is that you can test passes in isolation. Why did the mir tests did not reveal the bug? End-to-end tests are a SelectionDag feature.

They show the SUBREG_TO_REG, but SUBREG_TO_REG is often misunderstood and there's nothing that can check that is correctly used.

File: llvm/lib/Target/X86/GISel/X86InstructionSelector.cpp [hunk 1]
```diff
@@ -1778,12 +1778,9 @@ bool X86InstructionSelector::selectMulDivRem(MachineInstr &I,
         .addImm(8);
```
Comment (topperc) [inline llvm/lib/Target/X86/GISel/X86InstructionSelector.cpp:1778]: Probably for another patch, but we should match SelectionDAG here and use a MOVZX_NOREX or MOVSX_NOREX from AH here.

**Review by RKSimon — COMMENTED** [review]
Should we be losing the legalize tests (although maybe we should be merging 32/64 bit targets - no i64 divrem on x86 tests atm)? The select tests I don't find as much use for (and the isel-* tests are better to identify poor codegen IMO)

**Comment by arsenm** [comment]
> They show the SUBREG_TO_REG, but SUBREG_TO_REG is often misunderstood and there's nothing that can check that is correctly used.

We really should just remove SUBREG_TO_REG, it's impossible to use it correctly. My patch to hack around it in the coalescer is currently stuck on it's 5th or 6th revert due to unrelated issues it happens to hit

**Comment by e-kud** [comment]
I want to admit that it was too harsh to replace all `mir` tests with `ll` tests. Returned and unified legalizer tests as @RKSimon suggested. 

@tschuett do you think we still need updated select tests? These tests won't be much different from the existing ones that check invalid MIR. I think that the best check of instruction selection is valid final assembly. But I don't mind to keep them if we see their usefulness. Before fix it was an assert in register allocator due to `SUBREG_TO_REG` misuse.

**Comment by e-kud** [comment]
> They show the SUBREG_TO_REG, but SUBREG_TO_REG is often misunderstood and there's nothing that can check that is correctly used.

@topperc maybe we can add a subreg-superreg check to `MachineVerifier`?

File: llvm/lib/Target/X86/GISel/X86InstructionSelector.cpp (hunk 1 above)
Comment (e-kud) [reply llvm/lib/Target/X86/GISel/X86InstructionSelector.cpp:1778]: Do you mean to replace `COPY` + `SHR` by `movzx %ah, %ax`?

Comment (topperc) [reply llvm/lib/Target/X86/GISel/X86InstructionSelector.cpp:1778]: Yes this is the code from X86ISelDAGToDAG.cpp

```
    if (HiReg == X86::AH && !SDValue(Node, 1).use_empty()) {                     
      SDValue AHCopy = CurDAG->getRegister(X86::AH, MVT::i8);                    
      unsigned AHExtOpcode =                                                     
          isSigned ? X86::MOVSX32rr8_NOREX : X86::MOVZX32rr8_NOREX;              
                                                                                 
      SDNode *RNode = CurDAG->getMachineNode(AHExtOpcode, dl, MVT::i32,          
                                             MVT::Glue, AHCopy, InGlue);         
      SDValue Result(RNode, 0);                                                  
      InGlue = SDValue(RNode, 1);                                                
                                                                                 
      Result =                                                                   
          CurDAG->getTargetExtractSubreg(X86::sub_8bit, dl, MVT::i8, Result);    
                                                                                 
      ReplaceUses(SDValue(Node, 1), Result);                                     
      LLVM_DEBUG(dbgs() << "=> "; Result.getNode()->dump(CurDAG);                
                 dbgs() << '\n');                                                
    }
```

**Comment by arsenm** [comment]
> @topperc maybe we can add a subreg-superreg check to `MachineVerifier`?

Not really, the operation is more fundamentally broken. It is promising something about bits in registers which are not represented in the instruction

File: llvm/test/CodeGen/X86/GlobalISel/legalize-sdiv.mir [hunk 2]
```diff
@@ -0,0 +1,176 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py
+# RUN: llc -mtriple=x86_64-linux-gnu -run-pass=legalizer -verify-machineinstrs %s -o - | FileCheck %s --check-prefixes=CHECK,X64
+# RUN: llc -mtriple=i686-linux-gnu -run-pass=legalizer -verify-machineinstrs %s -o - | FileCheck %s --check-prefixes=CHECK,X86
+
+--- |
+  source_filename = "sdiv.ll"
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/GlobalISel/legalize-sdiv.mir:6]: Don't need the IR section

File: llvm/test/CodeGen/X86/GlobalISel/legalize-sdiv.mir [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    %2:_(s32) = G_SDIV %0, %1
+    $eax = COPY %2(s32)
+    RET 0, implicit $eax
+
+...
+---
+name:            test_sdiv_i64
+alignment:       16
+tracksRegLiveness: true
+registers:
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/GlobalISel/legalize-sdiv.mir:127]: Don't need registers or alignment sections

File: llvm/test/CodeGen/X86/isel-sdiv.ll [hunk 4]
```diff
@@ -0,0 +1,116 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc < %s -global-isel=0                    -mtriple=x86_64-linux-gnu | FileCheck %s --check-prefixes=X64
+; RUN: llc < %s -fast-isel -fast-isel-abort=1     -mtriple=x86_64-linux-gnu | FileCheck %s --check-prefixes=X64
+; RUN: llc < %s -global-isel -global-isel-abort=1 -mtriple=x86_64-linux-gnu | FileCheck %s --check-prefixes=X64
+; RUN: llc < %s -global-isel=0                    -mtriple=i686-linux-gnu | FileCheck %s --check-prefixes=X86,DAG-X86
+; RUN: llc < %s -fast-isel -fast-isel-abort=1     -mtriple=i686-linux-gnu | FileCheck %s --check-prefixes=X86,DAG-X86
+; RUN: llc < %s -global-isel -global-isel-abort=1 -mtriple=i686-linux-gnu | FileCheck %s --check-prefixes=X86,GISEL-X86
+
+define i8 @test_sdiv_i8(i8 %arg1, i8 %arg2) nounwind {
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/isel-sdiv.ll:9]: Maybe want vector tests somewhere too

**Review by arsenm — APPROVED** [review]
lgtm with nits

File: llvm/test/CodeGen/X86/isel-sdiv.ll (hunk 4 above)
Comment (RKSimon) [reply llvm/test/CodeGen/X86/isel-sdiv.ll:9]: FTR - I'd prefer vector tests to be in separate files as there is a lot of SSE/AVX level bloat

Comment (e-kud) [reply llvm/test/CodeGen/X86/isel-sdiv.ll:9]: @arsenm the vector work is in parallel with scalar instruction. We've started with some basics `G_BUILD_VECTOR` and `G_INSERT/EXTRACT_VECTOR_ELT`. We'll reiterate all operations ones more with adding tests for vector operands.

@RKSimon I agree with splitting vector/scalar cases, as well as float/scalar in some situations. I'm mostly motivated by creating a generic testsuite that can be reused for other targets when they start adding GlobalISel support incrementally.  I'm slightly struggling with existing mixed vector/float/int tests that I need to split for GlobalISel testing as not everything's implemented or to duplicate them.

However maybe this is not feasible as after some time there will be tests for non-standard sizes like `i42` that nobody will support at early stages.

File: llvm/test/CodeGen/X86/GlobalISel/legalize-sdiv.mir (hunk 2 above)
Comment (e-kud) [reply llvm/test/CodeGen/X86/GlobalISel/legalize-sdiv.mir:6]: I've always thought that IR section is needed to regenerate MIR if something has changed. Or is it not a common practice?

Comment (arsenm) [reply llvm/test/CodeGen/X86/GlobalISel/legalize-sdiv.mir:6]: No, it's needed for a specific set of IR references (block names, IR values in MMOs, and global references)

**Review by arsenm — APPROVED** [review]

**Pull request merged**